### ISSUE TITLE
Resource.asFile fails if the resource is in a JAR

### ISSUE THREAD

**Issue by argv-minus-one** [pathikrit/better-files#235]
I went to use `Resource.asFile`, and was faced with this:

	java.nio.file.FileSystemNotFoundException
		at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getFileSystem(ZipFileSystemProvider.java:169)
		at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getPath(ZipFileSystemProvider.java:155)
		at java.base/java.nio.file.Paths.get(Paths.java:143)
		at better.files.File$.apply(File.scala:1357)
		at better.files.File$.apply(File.scala:1354)

Some Googling revealed that it is *not* guaranteed that every valid resource URL (as in `java.net.URL`) is also a valid NIO URL (as in `java.nio.file.Path`). In particular, unlike with `java.net.URL`, handing a `jar` URL to `Paths.getPath` won't automatically mount the JAR file in question, and if it isn't mounted, then you get the above exception.

This issue is especially sneaky because it works fine for `file` URLs, like are used when running tests. Only when the app is bundled into a JAR does it fail. (That's why the `asFile` tests in `ResourceSpec` don't fail.)

As it stands, therefore, `Resource.asFile` is unsafe. It'll work for running tests, but won't work in production. I see the following options for dealing with this problem:

1. Remove `Resource.asFile`.

2. When the resource URL has the `jar` scheme, get the URL to the JAR file itself (that is, everything between `jar:` and `!/` in the URL), mount it if it isn't already mounted (using `java.nio.file.FileSystems#newFileSystem`), and unmount it when it's no longer needed.

	Problem: There does not appear to be any safe way to detect whether the new `FileSystem` is still in use, so it could potentially be closed while still in use, breaking whatever code was using it.

	Another problem: The JAR file's URL may not be supported by NIO, either. Applets, for instance, will break, because their resource URLs are like `jar:https://example.com/applet.jar!/com/example/app.conf`. Similarly, code might register its own custom `URLStreamHandler` and load classes from URLs with that custom scheme, but not register a corresponding `FileSystemProvider` for NIO.

3. Write a new stub `FileSystem` implementation that wraps `java.net.URL`.

Option 3 appears to be the only way to safely get a better-files `File` for a class resource. It also appears to be a ton of work! But `better.files.Resource` is pretty useless without it, so if we take option 1, we may as well remove `Resource` entirely.

What do we do? Has someone perhaps already written a `FileSystem` implementation as described in option 3? I can't seem to find such a thing on Google.

**Comment by pathikrit** [comment]
I think we should remove the `Resource.asFile` API completely. We should just keep 1 API i.e. `.url` (and everything else should be derived from it):

```scala
trait ResourceLoader {
  def url(name: String): Option[URL]
  def inputstream(name: String): Option[InputStream] = url(name)map(_.openStream())
  def asString(name: String): Option[String] = inputstream(name).map(_.asString)
} 
```

**Comment by pathikrit** [comment]
Fixed in #238
